enhancements/authentication/pod-security-admission-enforcement.md

```diff
@@ -0,0 +1,428 @@
+A new API is necessary to provide this flexibility.
+
+#### New API
+
+This API offers a gradual way to roll out Pod Security Admission enforcement to clusters.
+It gives users the ability to influence the rollout and see feedback on which Namespaces might violate Pod Security standards.
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PSAEnforcementMode indicates the actual enforcement state of Pod Security Admission
+// in the cluster. Unlike PSATargetMode, which reflects the user’s desired or “target”
+// setting, PSAEnforcementMode describes the effective mode currently active.
+//
+// The modes define a progression from no enforcement, through label-based enforcement
+// to label-based with global config enforcement.
+type PSAEnforcementMode string
+
+const (
+	// EnforcementModePrivileged indicates that no Pod Security restrictions
+	// are effectively applied.
+	// This aligns with a pre-rollout or fully "privileged" cluster state,
+	// where neither enforce labels are set nor the global config enforces "Restricted".
+	EnforcementModePrivileged PSAEnforcementMode = "Privileged"
+
+	// EnfrocementModeLabel indicates that the cluster is enforcing Pod Security
+	// labels at the Namespace level (via the PodSecurityAdmissionLabelSynchronizationController),
+	// but the global kube-apiserver configuration is still "Privileged."
+	EnfrocementModeLabel PSAEnforcementMode = "LabelEnforcement"
+
+	// EnforcmentModeFull indicates that the cluster is enforcing
+	// labels at the Namespace level, and the global configuration has been set
+	// to "Restricted" on the kube-apiserver.
+	// This represents full enforcement, where both Namespace labels and the global config
+	// enforce Pod Security Admission restrictions.
+	EnforcmentModeFull PSAEnforcementMode = "FullEnforcement"
+)
+
+// PSATargetMode reflects the user’s chosen (“target”) enforcement level.
+type PSATargetMode string
+
+const (
+	// TargetModePrivileged indicates that the user wants no Pod Security
+	// restrictions applied. The desired outcome is that the cluster remains
+	// in a fully privileged (pre-rollout) state, ignoring any label enforcement
+	// or global config changes.
+	TargetModePrivileged PSATargetMode = "Privileged"
+
+	// TargetModeConditional indicates that the user is willing to let the cluster
+	// automatically enforce a stricter enforcement once there are no violating Namespaces.
+	// If violations exist, the cluster stays in its previous state until those are resolved.
+	// This allows a gradual move towards label and global config enforcement without
+	// immediately breaking workloads that are not yet compliant.
+	TargetModeConditional PSATargetMode = "Conditional"
+
+	// TargetModeRestricted indicates that the user wants the strictest possible
+	// enforcement, causing the cluster to ignore any existing violations and
```

Comment:
Instead of ignoring existing violations, should the cluster operator go degraded with any existing violations being the reason? 

Maybe a knob for users to control this would be useful here to support use cases of:
- Forcefully enable `restricted` PSA configuration
- Enable `restricted` PSA configuration only if no violations (I realize this is what the `Conditional` enum type is for, but maybe that shouldn't be a separate option?)

Reply:
Note: I elaborated a bit more on my thinking here in https://github.com/openshift/enhancements/pull/1747#discussion_r1941778846